### PR TITLE
fix(lock): autodiscovery of lockfile

### PR DESCRIPTION
**Pull request by bartlomieju**
This commit adds autodiscovery of lockfile. 

This only happens if Deno discovers the configuration file (either 
"deno.json" or "deno.jsonc"). In such case Deno tries to load "deno.lock"
file that sits next to the configuration file, or creates one for user if
the lockfile doesn't exist yet.

As a consequence, "--lock" and "--lock-write" flags had been updated.
"--lock" no longer requires a value, if one is not provided, it defaults
to "./deno.lock" resolved from the current working directory. "--lock-write"
description was updated to say that it forces to overwrite a lockfile.

Autodiscovery is currently not handled by the LSP.

### CODE REVIEW

File: cli/args/flags.rs [hunk 1]
```diff
@@ -458,7 +458,9 @@ impl Flags {
       Some(files.clone())
     } else if let Run(RunFlags { script }) = &self.subcommand {
       if let Ok(module_specifier) = deno_core::resolve_url_or_path(script) {
-        if module_specifier.scheme() == "file" {
+        if module_specifier.scheme() == "file"
+          || module_specifier.scheme() == "npm"
```
Comment (bartlomieju) [inline cli/args/flags.rs:462]: This was fun - I tried this PR with Vite and couldn't get it to write to a lockfile. Turns out we never discover configuration file is the entrypoint is a `npm` package (eg. `deno run -A --unstable npm:vite`). This conditional now allows to discover configuration file in such case. Should I factor it out to a separate PR?

Comment (dsherret) [reply cli/args/flags.rs:462]: Oh, interesting. I think this makes sense to do.

**Review by ry — COMMENTED** [review]
Should we remove `--lock-write` flag?

What about a default value for `--lock`? 

If I change the lock file and then run again with `--lock=deno.json` I get this error:
```
# deno run --lock=deno.lock  --unstable -A  main.ts

============================================================
Deno has panicked. This is a bug in Deno. Please report this
at https://github.com/denoland/deno/issues/new.
If you can reliably reproduce this panic, include the
reproduction steps and re-run with the RUST_BACKTRACE=1 env
var set and include the backtrace in your report.

Platform: macos aarch64
Version: 1.27.0
Args: ["deno", "run", "--lock=deno.lock", "--unstable", "-A", "main.ts"]

thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', cli/npm/resolution.rs:311:43
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by bartlomieju** [comment]
> Should we remove --lock-write flag?

No, it allows to overwrite the lockfile completely - eg. when you get into a situation where there's integrity mismatch, but want to update the lockfile, then specifying this flag will do the job. Error message suggests it:

```
error: Integrity check failed for npm package: "cowsay@1.5.0".
  Cache has "sha512-8Ipzr54Z8zROr/62C8f0PdhQcDusS05gKTS87xxdji8VbWefWly0k8BwGK7+VqamOrkv3eGsCkPtvlHzrhWsCA==" and lockfile has "foobar".
  Use "--lock-write" flag to update the lockfile.
```

> What about a default value for --lock?

What would be the default value? `CWD/deno.lock`?

**Comment by ry** [comment]
> What would be the default value? CWD/deno.lock?

yes

> Can you provide contents of deno.lock and main.ts?

https://gist.github.com/ry/f313e2d8a56139e8ecb5fa7994d885e4

File: cli/args/flags.rs [hunk 2]
```diff
@@ -2145,16 +2147,17 @@ fn lock_arg<'a>() -> Arg<'a> {
   Arg::new("lock")
     .long("lock")
     .value_name("FILE")
-    .help("Check the specified lock file")
+    .help("Check the specified lock file. If not value is provided, defaults to deno.lock in the current working directory.")
```
Comment (dsherret) [inline cli/args/flags.rs:2150]:
```suggestion
    .help("Check the specified lock file. If value is not provided, defaults to deno.lock in the current working directory.")
```

**Comment by bartlomieju** [comment]
@ry all these problems have now been addressed.

File: cli/lockfile.rs [hunk 3]
```diff
@@ -82,35 +84,88 @@ pub struct Lockfile {
...
+    flags: &Flags,
+    maybe_config_file: Option<&ConfigFile>,
+  ) -> Result<Option<Lockfile>, AnyError> {
+    let filename = match flags.lock {
+      Some(ref lock) => PathBuf::from(lock),
+      None => match maybe_config_file {
+        Some(config_file) => {
+          if config_file.specifier.scheme() == "file" {
+            let mut path = config_file.specifier.to_file_path().unwrap();
+            path.set_file_name("deno.lock");
```
Comment (dsherret) [inline cli/lockfile.rs:97]: I was wondering if this should be `deno-lock.json` like `package-lock.json`, but I think this makes sense because then the formatter doesn't pick it up.

File: cli/lockfile.rs [hunk 4]
```diff
@@ -82,35 +84,88 @@ pub struct Lockfile {
...
+        content,
+        filename,
+      });
+    }
+
+    let result = match std::fs::read_to_string(&filename) {
+      Ok(content) => Ok(content),
+      Err(e) => {
+        if e.kind() == std::io::ErrorKind::NotFound {
+          let content = LockfileContent {
```
Comment (dsherret) [inline cli/lockfile.rs:132]: Maybe we should add something like `LockfileContent::empty()` to reduce duplication with the above.

File: cli/lsp/language_server.rs [hunk 5]
```diff
@@ -2916,6 +2916,8 @@ impl Inner {
         ..Default::default()
       },
       self.maybe_config_file.clone(),
+      // TODO(bartlomieju): add support for lockfile
```
Comment (dsherret) [inline cli/lsp/language_server.rs:2919]:
```suggestion
      // TODO(#16510): add support for lockfile
```

**Review by dsherret — APPROVED** [review]
LGTM! Will be exciting to have this.

File: cli/tests/integration/run_tests.rs [hunk 6]
```diff
@@ -3634,3 +3628,10 @@ fn websocket_server_idletimeout() {
 
   assert!(child.wait().unwrap().success());
 }
+
+itest!(auto_discover_lockfile {
+  args: "run run/auto_discover_lockfile/main.ts",
+  output: "run/auto_discover_lockfile/main.out",
+  http_server: true,
+  exit_code: 10,
```
Comment (dsherret) [inline cli/tests/integration/run_tests.rs:3636]: Seems like the exit code being `1` is failing the tests.

**Comment by bartlomieju** [comment]
> Is there a way to disable this behavior similar to how `--no-config` disables auto-discovery of config files?

Not currently, no. Is there a reason why you don't want to use a lockfile?
